funcx-faas/funcx-web-service
# Refactor tests to use fakeredis

This whole effort boils down to "where to mock". Trying to mock each and every use of a core class like `Task` is problematic. It's only exacerbated by that logic not being codified into a factory fixture.

The possibility of building mock data which is tailored to satisfy a test is a major reason for this effort. The fact that the tests were not robust to changes in the application was what originally motivated me.

I was hesitant to use `fakeredis` at first -- I didn't see it as necessary when we mostly want `hset`, `hget`, `lpush`, `lrange`, and a handful of other methods backed by a simple dict. However, after trying it out, it was an absolute pleasure to use. There are two compelling reasons I see to use this:
- it will support complex functionality like pubsub, which is harder than simple hstore to handle
- it's _someone else's_ implementation of dummy redis, so it's less likely that we'll suffer from confirmation bias and encode a mistake both into the application and into the tests

Change details follow.

---

- Add fakeredis to test requirements
- Use fakeredis client instead of a bare mock for `mock_redis`
- Create a mock_redis_task_factory fixture which creates tasks
- Update tests to use this and pass

It is no longer necessary to mock over the task initializer at all (the `from_id` method which is just a pointless copy of `__new__`).

Additional work which was necessary here:

Fix a bug that this surfaced: a missing call to `int()` on the invocation counter. Includes a test to confirm the failure behavior for non-int (bad) data.

Fix a bad test in testing task status -- 'result' and 'exception' get removed from task data when they are `None`.

For now, the pubsub mock is left in place, as it is rarely used. This should be changed to a pubsub backed by the same fakeserver at a later date.
